File: solutions/fortran/allergies/1/allergies.f90

module allergies
  implicit none

contains

  logical function allergicTo(allergy_str, allergy_key)
    character(len=*), intent(in) :: allergy_str
    integer, intent(in) :: allergy_key
    allergicTo = .false. 

    select case (allergy_str)
      case ("eggs")
        allergicTo = iand(1, allergy_key) == 1
      case ("peanuts")
        allergicTo = iand(2, allergy_key) == 2
      case ("shellfish")
        allergicTo = iand(4, allergy_key) == 4
      case ("strawberries")
        allergicTo = iand(8, allergy_key) == 8
      case ("tomatoes")
        allergicTo = iand(16, allergy_key) == 16
      case ("chocolate")
        allergicTo = iand(32, allergy_key) == 32
      case ("pollen")
        allergicTo = iand(64, allergy_key) == 64
      case ("cats")
        allergicTo = iand(128, allergy_key) == 128
    end select
  end function


  function allergicList(allergy_key)
    integer, intent(in) :: allergy_key
    character(len=100) :: allergicList
    allergicList = ''

    if (iand(1, allergy_key) == 1) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "eggs"
      else 
        allergicList(len_trim(allergicList)+1:) = "eggs"
      end if
    end if
    if (iand(2, allergy_key) == 2) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "peanuts"
      else 
        allergicList(len_trim(allergicList)+1:) = "peanuts"
      end if
    end if
    if (iand(4, allergy_key) == 4) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "shellfish"
      else 
        allergicList(len_trim(allergicList)+1:) = "shellfish"
      end if
    end if
    if (iand(8, allergy_key) == 8) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "strawberries"
      else 
        allergicList(len_trim(allergicList)+1:) = "strawberries"
      end if
    end if
    if (iand(16, allergy_key) == 16) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "tomatoes"
      else 
        allergicList(len_trim(allergicList)+1:) = "tomatoes"
      end if
    end if
    if (iand(32, allergy_key) == 32) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "chocolate"
      else 
        allergicList(len_trim(allergicList)+1:) = "chocolate"
      end if
    end if
    if (iand(64, allergy_key) == 64) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "pollen"
      else 
        allergicList(len_trim(allergicList)+1:) = "pollen"
      end if
    end if
    if (iand(128, allergy_key) == 128) then
      if (len_trim(allergicList) > 0) then
        allergicList(len_trim(allergicList)+2:) = "cats"
      else 
        allergicList(len_trim(allergicList)+1:) = "cats"
      end if
    end if
  end function
end module

File: solutions/cpp/allergies/1/allergies.cpp
#include "allergies.h"

#include <map>

namespace allergies
{
    /* 
     * Note: This could be potentially DRYed up via Boost Bimap, which allows
     * bidirectional use. However, I found the API a bit more challenging than
     * my patience allowed at the time of working on this!
     * https://www.boost.org/doc/libs/1_73_0/libs/bimap/doc/html/
     */
    const std::map<std::string, std::uint32_t> allergy_to_bitflag = {
        {"eggs", 0},
        {"peanuts", 1},
        {"shellfish", 2},
        {"strawberries", 3},
        {"tomatoes", 4},
        {"chocolate", 5},
        {"pollen", 6},
        {"cats", 7}};

    const std::map<std::uint32_t, std::string> bitflag_to_allergy = {
        {0, "eggs"},
        {1, "peanuts"},
        {2, "shellfish"},
        {3, "strawberries"},
        {4, "tomatoes"},
        {5, "chocolate"},
        {6, "pollen"},
        {7, "cats"}};

    allergy_test::allergy_test(std::uint32_t score)
    {
        allergies = score;
    }

    bool allergy_test::is_allergic_to(const std::string &allergy)
    {
        return allergies[allergy_to_bitflag.at(allergy)];
    }

    std::unordered_set<std::string> allergy_test::get_allergies()
    {
        std::unordered_set<std::string> allergies_set{};

        for (std::uint32_t i = 0; i < bitflag_to_allergy.size(); i++)
        {
            if (allergies[i])
            {
                allergies_set.insert(bitflag_to_allergy.at(i));
            }
        }

        return allergies_set;
    };

} // namespace allergies
